### PR TITLE
timedate, locale: fix barrier to suppress multiple function calls

### LINKED ISSUE #9672: timedated SetNtp: wrong PropertiesChanged value
**systemd version the issue has been seen with**
> 239

**Used distribution**
 > Debian unstable

When [testing Cockpit on Debian testing](https://github.com/cockpit-project/cockpit/pull/9700) we noticed a timedated/timesyncd regression, apparently introduced in systemd 239. Debian testing is the only distribution that has 239 - Fedora 28, RHEL etc. all still have 238 where this problem does not occur.

**Steps to reproduce the problem**

Try to disable and re-enable NTP:
```
busctl call org.freedesktop.timedate1 /org/freedesktop/timedate1 org.freedesktop.timedate1 SetNTP bb false false
busctl get-property org.freedesktop.timedate1 /org/freedesktop/timedate1 org.freedesktop.timedate1 NTP
sleep 1
busctl call org.freedesktop.timedate1 /org/freedesktop/timedate1 org.freedesktop.timedate1 SetNTP bb true false
busctl get-property org.freedesktop.timedate1 /org/freedesktop/timedate1 org.freedesktop.timedate1 NTP
```

One would expect the output "false" and then "true". Sometimes the second property get is false initially, and then true, i. e. it can take some time to settle; that's fine. However, sometimes (in about one out of three runs for me) it doesn't become true even after several minutes.

When that happens, the D-Bus property is false, but timesyncd is running:
```
● systemd-timesyncd.service - Network Time Synchronization
   Loaded: loaded (/lib/systemd/system/systemd-timesyncd.service; enabled; vendor preset: enabled)
   Active: active (running) since Fri 2018-07-20 11:35:17 EDT; 39s ago
     Docs: man:systemd-timesyncd.service(8)
 Main PID: 8299 (systemd-timesyn)
   Status: "Synchronized to time server 193.141.27.1:123 (0.debian.pool.ntp.org)."
    Tasks: 2 (limit: 1152)
   Memory: 948.0K
   CGroup: /system.slice/systemd-timesyncd.service
           └─8299 /lib/systemd/systemd-timesyncd

Jul 20 11:35:16 unassigned-hostname systemd[1]: Starting Network Time Synchronization...
Jul 20 11:35:17 unassigned-hostname systemd[1]: Started Network Time Synchronization.
Jul 20 11:35:17 unassigned-hostname systemd-timesyncd[8299]: Synchronized to time server 193.141.27.1:123 (0.debian.pool.ntp.org).
```

Interestingly, when I call `timedatectl` in this steady broken state, it initially agrees ("NTP service: inactive"), but a second or so later both the D-Bus property and the output of `timedatectl` NTP service: become `true`. So calling `timedatectl` somehow "tickles" timedated to update the property.

### PR DESCRIPTION
**Pull request by yuwata**
Fixes #9672.

### CODE REVIEW

**Comment by martinpitt** [comment]
Thanks @yutawa! With this PR applied, I did not get the "NTP flag is eternally wrong" bug any more. It is stil `false` right after enabling NTP (which is a bit confusing), and our entire integration test still fails, but this is something to debug on my side, at least for now. On the CLI this definitively helps.

**Comment by yuwata** [comment]
@martinpitt Thank you for checking this PR. I've force-pushed an updated version. Now, timedated wait after stopping/starting NTP client service. PTAL.

File: src/timedate/timedated.c [hunk 1]
```diff
@@ -846,9 +916,11 @@ static int method_set_ntp(sd_bus_message *m, void *userdata, sd_bus_error *error
 
         log_info("Set NTP to %sd", enable_disable(enable));
 
-        (void) sd_bus_emit_properties_changed(bus, "/org/freedesktop/timedate1", "org.freedesktop.timedate1", "NTP", NULL);
+        r = sd_bus_message_new_method_return(m, &c->reply_set_ntp);
+        if (r < 0)
+                return r;
```
Comment (keszybz) [inline src/timedate/timedated.c]: Hmm, this changes the semantics of `method_set_ntp` significantly — before it'd be "set the state" and now it is "set the state, wait for the unit to be started or stopped". Additional wrinkle is that if a second parallel `method_set_ntp` is made in parallel, the first one will cut the wait short, so the condition is really more like "wait for the unit to be started or stopped OR for another method_set_ntp call to come in". Shouldn't the change be made smaller, to keep the call asynchronous, and only wait on the job to be finished to send out the property change? This should fix the original issue, without changing the semantics of the call. I'm not sure what the proper solution is, so I'm putting this out here for discussion.

Comment (martinpitt) [reply src/timedate/timedated.c]: I confirm that this fixes the wrong property changed value. However, I agree with @keszybz that it seems cleaner to not make `SetNTP` blocking, and just defer the `PropertyChanged` signal until the NTP unit is actually running.

Comment (yuwata) [reply src/timedate/timedated.c]: Ah, I see. OK, I will update this. Thank you.

**Comment by yuwata** [comment]
@martinpitt and @keszybz Updated. Changes are
- Sending reply is not delayed now. Only emitting property change is delayed.
- When the previous job of starting/stopping unit has not completed, then refuse subsequent SetNTP method calls.

PTAL. Thank you.

**Comment by yuwata** [comment]
@martinpitt Thank you for testing this. I've updated commit message.

**Comment by keszybz** [comment]
> When the previous job of starting/stopping unit has not completed, then refuse subsequent SetNTP method calls.

No, please don't do that. As you write in the commit message, this method can be called frequently. Now that would become an error and require special handling. Instead this should behave the similarly to repeated `systemctl start/stop` calls. When a new start/stop job is queued, the old one will be cancelled, so the latest one wins. I think it should be enough to just watch for the latest queued job, and forget the earlier ones.

**Comment by martinpitt** [comment]
> I think it should be enough to just watch for the latest queued job, and forget the earlier ones.

Oh, that explains a strange new cockpit test case failure that I've seen with this PR applied. Indeed!

**Comment by yuwata** [comment]
@keszybz and @martinpitt Thank you for the comment. I've force-pushed a revised version. Changes from the previous revision are the following:
- drop the previously introduced bus error.
- forget uncompleted previous job.

File: src/timedate/timedated.c [hunk 2]
```diff
@@ -804,6 +860,11 @@ static int method_set_ntp(sd_bus_message *m, void *userdata, sd_bus_error *error
         if (r == 0)
                 return 1;
 
+        /* This method may be called frequently. Forget the previous job if it has not completed yet.
+         * In that case, the active state obtained by context_update_ntp_status() may no be reliable. */
```
Comment (keszybz) [inline src/timedate/timedated.c]: Typo: no

File: src/timedate/timedated.c [hunk 3]
```diff
@@ -800,6 +860,11 @@ static int method_set_ntp(sd_bus_message *m, void *userdata, sd_bus_error *error
...
+        ignore_active_state = c->slot_job_removed;
+        c->slot_job_removed = sd_bus_slot_unref(c->slot_job_removed);
```
Comment (keszybz) [inline src/timedate/timedated.c]: So... is this check really necessary? Wouldn't it be simpler and more reliable to just send the job start/stop request always? Imagine the following scenario:
1. the service is not running
2. somebody calls enable through `timedated`, `timedated` queues a start job
3. the start job finishes
4. a stop job is queued through `systemctl stop`
5. somebody calls enable through `timedated`, `timedated` queues a start job

Even if 4 and 5 are always in that order, depending on whether 3 is earlier or later, the outcome will be different (in one case a new start job will be queued, in the other not).

My thinking is that pid1 is the only place that really know the current state, and the only cost we have from forwarding all the enable/disable calls to it is a dbus call, which is not expensive.

Comment (yuwata) [reply src/timedate/timedated.c]: That is a very good question. While I update this, I noticed that. So, can I confirm your suggestion? You mean that we should set in some sense that `ignore_active_state` always true?

I will test the racy conditions later and may update this. Thank you so much for giving comments so many times.

**Comment by yuwata** [comment]
@keszybz Updated.
- Always call StartUnit/StopUnit regardless its active state.

**Comment by keszybz** [comment]
@yuwata thanks.

**Pull request merged**